### ISSUE TITLE
Breaking change in nested autocmds in #23256

### ISSUE THREAD

**Issue by ii14** [neovim/neovim#23368]
#23256 introduced a subtle breaking change.

When triggering an event, the last autocmd is marked as such and the execution of autocmds
stops once it's reached, to prevent infinite loops when creating new autocmds from inside
autocmds. It can be thought of as saving current autocmds so they can't be changed during
their execution (they can be still deleted though).

Before #23256 last autocmd was marked globally. This means that nested event can add new
autocmds to the event that triggered it, during its execution. #23256 fixes that by
keeping track of the last autocmd per executed event. It is a breaking change though and
reportedly broke some code.

```lua
local skip = false
local level = 0

vim.api.nvim_create_autocmd('FileType', {
  pattern = 'test',
  callback = function()
    level = level + 1
    print(('  '):rep(level) .. 'autocmd 1')

    -- only define autocmd 2 and trigger the event once
    if not skip then
      skip = true

      vim.api.nvim_create_autocmd('FileType', {
        pattern = 'test',
        callback = function()
          level = level + 1
          print(('  '):rep(level) .. 'autocmd 2')
          level = level - 1
        end,
      })

      print(('  '):rep(level) .. 'EVENT 2')
      vim.api.nvim_exec_autocmds('FileType', { pattern = 'test', modeline = false })
    end

    level = level - 1
  end,
})

print(('  '):rep(level) .. 'EVENT 1')
vim.api.nvim_exec_autocmds('FileType', { pattern = 'test', modeline = false })
```

Annotated output before #23256:
```
EVENT 1         - internally sets autocmd 1 as the last autocmd
  autocmd 1     - creates autocmd 2
  EVENT 2       - internally sets autocmd 2 as the last autocmd
    autocmd 1   - skipped
    autocmd 2
  autocmd 2     - autocmd 2 is now executed by event 1. wouldn't happen if event 2 was not triggered
```

Annotated output after #23256:
```
EVENT 1         - internally sets autocmd 1 as the last autocmd for event 1 only
  autocmd 1     - creates autocmd 2
  EVENT 2       - internally sets autocmd 2 as the last autocmd for event 2 only, doesn't affect event 1 at all
    autocmd 1   - skipped
    autocmd 2
```

What do we do about it? I think the new behavior is better and that seemed to be the
intention in the first place, but it is a breaking change.

**Comment by clason** [comment]
Do you have a link to an issue about this breaking a plugin in the wild (so we can estimate the actual impact)?

**Comment by clason** [comment]
Yes. You should not lazy-load filetype plugins, period. (I mean, "you do you, whatever sparks joy for you" and all that, but that's not something we need to officially support.)

**Comment by ii14** [comment]
There is another inconsistent behavior that was fixed. Before #23256 autocmd 3 was
executed only if autocmd 2 is defined and autocmd 1 has a different pattern than 2 and 3.
This doesn't make any sense whatsoever unless you're familiar with the implementation:

```lua
local pat1 = 'test'
local pat2 = '*'

vim.api.nvim_create_autocmd('User', {
  pattern = pat1,
  callback = function()
    print('autocmd 1')

    vim.api.nvim_create_autocmd('User', {
      pattern = pat2, -- different pattern!
      callback = function() print('autocmd 3') end,
    })
  end,
})

vim.api.nvim_create_autocmd('User', {
  pattern = pat2, -- different pattern!
  callback = function() print('autocmd 2') end,
})

vim.api.nvim_exec_autocmds('User', { pattern = 'test', modeline = false })
```

Before:
```
autocmd 1
autocmd 2
autocmd 3
```
After #23256 or if `local pat2 = pat1` (!):
```
autocmd 1
autocmd 2
```

**Comment by justinmk** [comment]
It's good that this is documented here, but looks like `wontfix`. Is there a 2-line explanation we can mention in `:help nvim-changed` ?

**Comment by ii14** [comment]
Not sure about short description, can we just say something like "fixed inconsistent behavior in nested autocommands" and link to this issue?
